storage/storage_image.go

```diff
@@ -0,0 +1,430 @@
+	if err != nil && err != io.EOF {
+		return "", -1, err
+	}
+	// Set up to read the whole blob (the initial snippet, plus the rest)
+	// while digesting it with sha256.
+	hasher := sha256.New()
+	defragmented := io.MultiReader(bytes.NewBuffer(header[:n]), stream)
+	multi := io.TeeReader(defragmented, hasher)
+	if (n > 0) && archive.IsArchive(header[:n]) {
+		// It's a filesystem layer.  If it's not the first one in the
+		// image, we assume that the most recently added layer is its
+		// parent.
+		parentLayer := ""
+		if len(s.BlobList) > 0 {
+			for _, blob := range s.BlobList {
+				if layerID, ok := s.Layers[blob]; ok {
+					parentLayer = layerID
+				}
+			}
+		}
+		// Now try to figure out if the identifier we have is an ID or
+		// something else, so that we can do collision detection
+		// correctly.
+		names := []string{digest}
+		id := digest
+		if matches := idRegexp.FindStringSubmatch(digest); len(matches) > 1 {
+			id = matches[len(matches)-1]
+			names = nil
+		}
+		layer, err := s.store.CreateLayer(id, parentLayer, names, "", true)
+		if err != nil && err != storage.ErrDuplicateID {
+			return "", -1, err
+		}
+		if err == storage.ErrDuplicateID {
+			// There's already a layer with the same ID.  If its
+			// checksum is the same as ours, then we're cool with
+			// just using the already-stored copy.
+			io.Copy(ioutil.Discard, multi)
+			hash := hasher.Sum(nil)
+			hasher = sha256.New()
+			diff, _, err := s.diffLayer(id, false)
+			if err != nil {
+				logrus.Debugf("blob %q already present but got %v while verifying it", err)
+				return "", -1, err
+			}
+			io.Copy(hasher, diff)
+			diff.Close()
+			if bytes.Compare(hash, hasher.Sum(nil)) != 0 {
+				// They are not the same.  Return an error.
+				logrus.Debugf("blob %q already present with different contents", id)
+				return "", -1, storage.ErrDuplicateID
+			}
+			// They are the same.  Record that this blob is a layer.
+			logrus.Debugf("blob %q already present with same contents, continuing", id)
+			if layer, err = s.store.GetLayer(id); err == nil && layer != nil {
+				id = layer.ID
+			}
+			s.Layers[digest] = id
+		} else {
+			// Apply the layer.
+			actualSize, err = s.store.ApplyDiff(layer.ID, multi)
+			if err != nil {
+				s.store.DeleteLayer(layer.ID)
+				return "", -1, err
+			}
+			// If the diff was compressed, make a note of how it
+			// was compressed, so that we can compress it again if
+			// we need to extract it via GetBlob().
+			layerdata := storageLayerMetadata{
+				Compression:  archive.DetectCompression(header[:n]),
+				ExpectedSize: expectedSize,
+			}
+			if metadata, err := json.Marshal(layerdata); len(metadata) != 0 && err == nil {
+				s.store.SetMetadata(layer.ID, string(metadata))
+			}
+			// Record that this blob is a layer.
+			s.Layers[digest] = layer.ID
+			logrus.Debugf("blob %q imported as a filesystem layer", id)
+		}
+		// Check that our computed sha256sum matches the digest, then
+		// add it to the blob list.
```

Comment:
Most importantly, somewhere in the `oci` changes `digest` became a value which may not be provided if unknown on input; this `PutBlob` must be able to handle that.

Otherwise, `copy.Image` does verify that the digest, if provided, matches the input stream, and it is necessary only to read until EOF and `copy.Image` will raise the error; so computing the digest _again_ in here just to detect that is unnecessary for `copy.Image`, and signature verification. (Though, if containers/storage, for its own purposes, requires digests to match, it is of course free to do any verification/enforcement it needs).

For signing etc., it is important not to _use_ the input stream if it should match a digest and doesn’t; but it is less important to _report_ that the stream doesn’t match. Look at how `dockerImageDestination.PutBlob` checks when, if a digest is provided, the registry claims to store it; in that case it discards the input stream without reading any of it, and the stream is never verified against the digest—but that is fine, because anyone downloading from the registry will verify against the provided digest, and we don't care at all how the registry got the correct blob.

OTOH, when we _are_ using the stream, `PutBlob` needs to implement computing the digest to have something to return when the digest is not provided on input, and in that case it is simpler to compute it always than to have a conditional in there. And it is unnecessary for `PutBlob` to compare the computed digest with the provided one, `copy.go` does that / will to that (see my `compress` branch).

It seems that the `dockerImageDestination.PutBlob` case is very similar to what is happening here with `storage.ErrDuplicateID`.